File: python/machine_learning/k_nearest_neighbor.py
from sklearn.neighbors import NearestNeighbors
import numpy as np


def knn(X, Y, K):
    """Find the K nearest neighbor in X for each point in Y.
    Parameters:
        X (numpy array): A matrix, where the columns are the features and the rows are the examples.
        Y (numpy array): A matrix to query, where the columns are the features and the rows are the examples.
    Returns:
        distances (numpy array): Distance of the K nearest neighbor in X to Y.
        indices (numpy array): Indexes of the K nearest neighbor in X to Y.
    Examples:
        >>> np.random.seed(0)
        >>> X = np.random.randn(100000,512)
        >>> Y = np.random.randn(2,512)
        >>> dist, idx = knn(X,Y,5) #takes 5.78s
        >>> idx.shape
        (2, 5)
        >>> dist.shape
        (2, 5)

    """
    nbrs = NearestNeighbors(n_neighbors=K, algorithm='auto', metric='euclidean', n_jobs=-1).fit(X)
    distances, indices = nbrs.kneighbors(Y)
    return distances, indices

File: matlab/machine_learning/k_nearest_neighbor.m
% K nearest neighbor
%
% knnsearch finds the nearest neighbor in X for each point in Y.
% IDX is the index of the K nearest neighbors in X. 
% D is the distance of the K nearest neighbor in X to Y.
% IDX, D and Y have the same number of rows.
% source: https://es.mathworks.com/help/stats/knnsearch.html

rng('default') % for reproducibility
X = randn(100000,512);
Y = randn(2,512);
tic
[IDX,D] = knnsearch(X,Y,'K',5,'Distance','euclidean')
toc
